lib/recursive-methods.rb

```diff
@@ -1,49 +1,95 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError, "Please pick positive integer" if n < 0
+  return 1 if n == 0
+  n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  s[-1] + reverse(s[0...-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
```

Comment:
⚠️ This is done in-place, but it's not recursive.  Can you see any ways to do this using the following method header

`def reverse_inplace(s, i = 0, j = s.length - 1)`

Reply:
Thanks for the hint! How about now?